File: packages/Teams/TriggerConditions/OnTeamsMEFetchTask.cs
﻿// Licensed under the MIT License.
// Copyright (c) Microsoft Corporation. All rights reserved.

using System.Collections.Generic;
using System.Runtime.CompilerServices;
using AdaptiveExpressions;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Adaptive.Conditions;
using Microsoft.Bot.Connector;
using Newtonsoft.Json;

namespace Microsoft.Bot.Components.Teams.Conditions
{
    /// <summary>
    /// Actions triggered when a Teams InvokeActivity is received when activity.name='composeExtension/fetchTask'.
    /// </summary>
    public class OnTeamsMEFetchTask : OnInvokeActivity
    {
        [JsonProperty("$kind")]
        public new const string Kind = "Teams.OnMEFetchTask";

        [JsonProperty("commandId")]
        public string CommandId { get; set; }

        [JsonConstructor]
        public OnTeamsMEFetchTask(string commandId = null, List<Dialog> actions = null, string condition = null, [CallerFilePath] string callerPath = "", [CallerLineNumber] int callerLine = 0)
            : base(actions: actions, condition: condition, callerPath: callerPath, callerLine: callerLine)
        {
            CommandId = commandId;
        }

        /// <inheritdoc/>
        protected override Expression CreateExpression()
        {
            // if name is 'composeExtension/fetchTask'
            var expressions = new List<Expression>
            {
                Expression.Parse($"{TurnPath.Activity}.ChannelId == '{Channels.Msteams}' && {TurnPath.Activity}.name == 'composeExtension/fetchTask'"),
                base.CreateExpression()
            };

            if (!string.IsNullOrEmpty(CommandId))
            {
                expressions.Add(Expression.Parse($"{TurnPath.Activity}.value.commandId == '{CommandId}'"));
            }

            return Expression.AndExpression(expressions.ToArray());
        }
    }
}

File: packages/Teams/js/src/conditions/onTeamsMEFetchTask.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { Expression } from 'adaptive-expressions';
import { Channels } from 'botbuilder';
import { TurnPath } from 'botbuilder-dialogs';
import { OnInvokeActivity } from 'botbuilder-dialogs-adaptive';

/**
 * Actions triggered when a Teams InvokeActivity is received when activity.name='composeExtension/fetchTask'.
 */
export class OnTeamsMEFetchTask extends OnInvokeActivity {
  static $kind = 'Teams.OnMEFetchTask';

  public commandId?: string;

  /**
   * Create expression for this condition.
   *
   * @returns {Expression} An [Expression](xref:adaptive-expressions.Expression) used to evaluate this rule.
   */
  protected createExpression(): Expression {
    const expressions = [
      Expression.parse(
        `${TurnPath.activity}.channelId == '${Channels.Msteams}' && ${TurnPath.activity}.name == 'composeExtension/fetchTask'`
      ),
      super.createExpression(),
    ];

    if (this.commandId) {
      expressions.push(
        Expression.parse(
          `${TurnPath.activity}.value.commandId == '${this.commandId}'`
        )
      );
    }

    return Expression.andExpression(...expressions);
  }
}
